### PR TITLE
[ci] Add linter for PR title and body

### PR DESCRIPTION
**Pull request by driazati**
This adds a lint step that checks the PR title and body with some of the basic checks from https://github.com/apache/tvm-rfcs/pull/88. This makes it so we can enforce the decided rules without having to rely on automated fix-ups, though for certain things those would be ideal in the future. The rules implemented here are not complete as the RFC is not fully merged but this is just supposed to be a starting point with some of the basic rules, namely that a title and body should be present. A follow up PR will enforce that there are no @ tags once #12637 is resolved.

cc @Mousius @areusch @gigiblender @gromero @leandron

### CODE REVIEW

File: tests/scripts/check_pr.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+class Check:
+    # check to run, returning OK means it passed, anything else means it failed
+    check: Callable[[str], Any]
+
+    # function to call to generate the error message
+    error_fn: Callable[[Any], str]
+
+
+def non_empty(s: str):
+    if len(s) == 0:
```
Comment (areusch) [inline tests/scripts/check_pr.py:44]: should we s.trim()?

File: tests/scripts/check_pr.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        return False
+    return OK
+
+
+def usernames(s: str):
+    m = GITHUB_USERNAME_REGEX.findall(s)
+    if m and len(m) > 0:
```
Comment (areusch) [inline tests/scripts/check_pr.py:51]: i think len(m) check isn't needed

File: tests/scripts/check_pr.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+from dataclasses import dataclass
+from typing import Any, List, Callable
+
+
+from git_utils import GitHubRepo, parse_remote, git
+from cmd_utils import init_log, tags_from_title
+
+
+GITHUB_USERNAME_REGEX = re.compile(r"(@[a-zA-Z0-9-]+)", flags=re.MULTILINE)
+OK = object()
```
Comment (areusch) [inline tests/scripts/check_pr.py:31]: :clap:

File: tests/scripts/check_pr.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    # check to run, returning OK means it passed, anything else means it failed
+    check: Callable[[str], Any]
+
+    # function to call to generate the error message
+    error_fn: Callable[[Any], str]
+
+
+def non_empty(s: str):
+    if len(s) == 0:
+        return False
```
Comment (areusch) [inline tests/scripts/check_pr.py:45]: could consider just adding a dedicated `FAIL` here too

File: tests/scripts/check_pr.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+    if args.pr_body:
+        body = args.pr_body
+        title = args.pr_title
+    else:
+        remote = git(["config", "--get", f"remote.{args.remote}.url"])
+        user, repo = parse_remote(remote)
+
+        github = GitHubRepo(token=os.environ["GITHUB_TOKEN"], user=user, repo=repo)
+        pr = github.get(f"pulls/{args.pr}")
+        body = pr["body"]
```
Comment (areusch) [inline tests/scripts/check_pr.py:139]: wondering if we should just pass `pr` to a check function and not have title/body separately?

**Review by areusch — COMMENTED** [review]
awesome!

File: tests/scripts/check_pr.py (hunk 5 above)
Comment (driazati) [reply tests/scripts/check_pr.py:139]: doing it this way lets things get re-used more easily across both (like non-empty check, not a huge deal for this code as-is but if we add more checks later it'd become more apparent)

**Review by areusch — APPROVED** [review]

**Comment by vinx13** [comment]
hey @driazati thanks for adding this linter. I found it not throwing meaningful CI failure message when PR body is missing. Here are the CI error message:
```
[2022-09-01T20:59:13.733Z] Traceback (most recent call last):
[2022-09-01T20:59:13.733Z]   File "ci/scripts/check_pr.py", line 136, in <module>
[2022-09-01T20:59:13.733Z]     body = body.strip()
[2022-09-01T20:59:13.733Z] AttributeError: 'NoneType' object has no attribute 'strip'
```
Shall we improve it?

**Comment by driazati** [comment]
thanks for flagging that! #12680 should fix it

**Pull request merged**